### PR TITLE
upgrade ember-cli to 3.28.5; make compatible with Ember 4.x

### PR DESCRIPTION
**Pull request by Gaurav0**
Breaking change: Implicit Injections are removed (since they are deprecated in 3.x and removed in 4.x). Services such as `session` and `torii` must be injected explicitly into routes and controllers like this:

```
import Route from '@ember/routing/route';
import AuthenticatedRouteMixin from 'torii/routing/authenticated-route-mixin';
import { inject as service } from '@ember/service';

export default class MyAuthenticatedRoute extends Route.extend(AuthenticatedRouteMixin) {
  @service session; // explicit injection
}
```

Tests run here: https://github.com/Gaurav0/torii/actions

### CODE REVIEW

**Comment by halfbyte** [comment]
Hi @Gaurav0 thanks for your work on this and your patience, I've been on a looong summer break. 

I'm not super happy with the size and the shape of this PR, my plan was to make the tests run on Ember 3 LTS and then gradually do the changes for Ember 4 - If you have it all figured out already it would be dumb to waste your work, though - Let me have a longer look at this in a more wake state than I am in right now.

**Comment by Gaurav0** [comment]
@halfbyte I used @jmonster 's previous work which already had tests running on Ember 3 LTS. There's a lot of work sitting around that never got merged. :)

**Comment by Gaurav0** [comment]
@halfbyte SeeClickFix has been using this in production now, with thousands of users.

Let me know if I can help get torii updated.

**Comment by jmonster** [comment]
@Gaurav0 my PR is/was being used by Heroku’s dashboard as of about a year ago. We gave up on getting the PR merged but good luck to you!

**Comment by halfbyte** [comment]
@Gaurav0 @jmonster Thanks for your input. I think what I'll do is I'll try to compare the work me and @mpminardi did sort of file by file next week and then probably merge in this one either unmodified and do a few corrections on top or maybe come back to you @Gaurav0 and ask for some changes if I think they'll be needed before a merge.

I'll need to figure out where and how to publish a new version and see if we'll need to add the `@namespace` @Gaurav0 rightfully complained about in #5 - I guess it's a matter of figuring out if we can contact someone with the original access rights. 

I swear I am not making this complicated on purpose, but I do feel a bit of pressure suddenly having inherited the future of this project.

Thanks for your work - We'll make it all work somehow, I'm sure.

**Comment by Gaurav0** [comment]
@halfbyte , @mixonic and @coryforsythe should have any access rights for npm publishing you are looking for.

File: tests/unit/services/popup-test.js [hunk 1]
```diff
@@ -1,180 +1,237 @@
...
+  test('open rejects when window closes', function (assert) {
+    let done = assert.async();
+
+    const mockWindow = buildMockWindow();
+    window.open = function () {
+      assert.ok(true, 'calls window.open');
+      return mockWindow;
+    };
+
+    run(function () {
```
Comment (halfbyte) [inline tests/unit/services/popup-test.js:195]: Looking at my own branch, I think these `run()` wrappers are not needed anymore?

**Review by halfbyte — COMMENTED** [review]
@Gaurav0 Well, that was one heck of a PR to review. Gonna get me a ☕ now. 

I have left two comments, more questions than actual change requests.

All in all this looks quite good and mergeable. 

I have asked Matthew Beale (Mixonic) about the npm access stuff - For now I think I'll revert the namespacing on the main branch.

File: app/initializers/initialize-torii.js [hunk 2]
```diff
@@ -1,18 +1,20 @@
...
   initialize(application) {
-    if (arguments[1]) { // Ember < 2.1
+    if (arguments[1]) {
+      // Ember < 2.1
       application = arguments[1];
     }
+
     configure(config.torii || {});
     bootstrapTorii(application);
     application.inject('route', 'torii', 'service:torii');
```
[the remark this answers is not in the document]
Comment (Gaurav0) [reply app/initializers/initialize-torii.js:16]: Yes, it is non-functional in Ember 4.x. I think it should be removed as well. I'll update.

File: tests/unit/services/popup-test.js (hunk 1 above)
Comment (Gaurav0) [reply tests/unit/services/popup-test.js:195]: You are probably correct. I'll try it without them.

**Pull request merged**